# Lab Assignment 9: Data Management Using `pandas`, Part 2
## DS 6001: Practice and Application of Data Science
## Name: Afnan Alabdulwahab

### Instructions
Please answer the following questions as completely as possible using text, code, and the results of code as needed. Format your answers in a Jupyter notebook. To receive full credit, make sure you address every part of the problem, and make sure your document is formatted in a clean and professional way.

In this lab, we are going to build the Country Analysis Relational DataBase (which we will call the C.A.R.D.B. or the "Cardi B"):

![CardbiB](https://media.giphy.com/media/3oEjI5ry4IwZ3RDw6k/giphy.gif "cardib")

We will be collecting data from two sources. First, we will use open data from the World Bank's [Sovereign
Environmental, Social, and Governance (ESG) Data](https://datatopics.worldbank.org/esg/) project. The ESG data reports data from every country in the world over the time frame from 1960-2022 on a wide variety of topics including education, health, and economic factors within the countries. Second, we will use data on the quality and democratic character of countries' governments as reported by the [Varieties of Democracy (V-Dem)](https://www.v-dem.net/data/the-v-dem-dataset/) project at the University of Notre Dame. By using both data sources, we can conduct analyses to see whether democratic openness leads to better societal outcomes for countries. We can also write queries to capture a wide range of information on countries' political parties, tax systems, and banking industries, for example. Or as Cardi B would say, "You in the club just to party, I'm there, I get paid a fee. I be in and out them banks so much, I know they're tired of me."

## Problem 0
Import the following packages (use `pip install` to download any packages you don't already have installed):

In [1]:
import numpy as np
import pandas as pd
import requests
import os
import io
import zipfile

Both the World Bank and V-Dem store their data in zipped directories containing CSV files. Download the World Bank data into your current working directory by typing the following code:

And download the V-Dem data by typing:

After you've run this code successfully once, the files you need will be in your working directory and you should save time by switching these cells from "code" to "raw" so that they don't run again if you restart the kernel.

You will only need three of the files you've downloaded. Load the 'V-Dem-CY-Core-v13.csv' file as `vdem` and the 'ESGData.csv' file as `wb`. 

In [2]:
vdem = pd.read_csv('V-Dem-CY-Core-v13.csv')
wb = pd.read_csv('ESGCSV.csv')

## Problem 1
First, let's focus on the `vdem` data ('V-Dem-CY-Core-v13.csv'). Use `pandas` methods to perform the following tasks:

### Part a
Keep only the 'country_text_id', 'country_name','year', 'v2x_polyarchy', and 'v2peedueq' columns. [1 point]

In [3]:
# Setting pandas option to display all columns
pd.set_option('display.max_columns', None) 

To include only these columns I am defining a list of these column names, then passing the list to the dataframe index as follows:

In [4]:
vdem = vdem[['country_text_id', 'country_name', 'year', 'v2x_polyarchy', 'v2peedueq' ]]
vdem

country_text_id       country_name  year  v2x_polyarchy  v2peedueq
0                 MEX             Mexico  1789          0.028        NaN
1                 MEX             Mexico  1790          0.028        NaN
2                 MEX             Mexico  1791          0.028        NaN
3                 MEX             Mexico  1792          0.028        NaN
4                 MEX             Mexico  1793          0.028        NaN
...               ...                ...   ...            ...        ...
27550             SPD  Piedmont-Sardinia  1857          0.207        NaN
27551             SPD  Piedmont-Sardinia  1858          0.210        NaN
27552             SPD  Piedmont-Sardinia  1859          0.210        NaN
27553             SPD  Piedmont-Sardinia  1860          0.213        NaN
27554             SPD  Piedmont-Sardinia  1861          0.213        NaN

[27555 rows x 5 columns]

### Part b
Use the `.query()` method to keep only the rows in which year is greater than or equal to 1960 and less than or equal to 2021. [1 point]

Using pandas' `.query()` method with a logical condition `'year >= 1960 & year <= 2021'` then the result is reassigned to `vdem`:

In [5]:
vdem = vdem.query('year >= 1960 & year <= 2021')
vdem

country_text_id country_name  year  v2x_polyarchy  v2peedueq
171               MEX       Mexico  1960          0.232     -1.438
172               MEX       Mexico  1961          0.234     -1.438
173               MEX       Mexico  1962          0.233     -1.438
174               MEX       Mexico  1963          0.233     -1.438
175               MEX       Mexico  1964          0.231     -1.438
...               ...          ...   ...            ...        ...
26150             ZZB     Zanzibar  2017          0.267      1.661
26151             ZZB     Zanzibar  2018          0.268      1.486
26152             ZZB     Zanzibar  2019          0.266      1.486
26153             ZZB     Zanzibar  2020          0.258      1.427
26154             ZZB     Zanzibar  2021          0.276      1.779

[10371 rows x 5 columns]

### Part c
Rename 'country_text_id' to 'country_code', 'country_name' to 'country_name_vdem', 'v2x_polyarchy' to 'democracy', and 'v2peedueq' to 'educational_equality'. [1 point]

Using the `.rename()` method on the dataframe, `vdem`, with a dictionary parameter that contains mapping of the old column names to the new names, and `axis=1` to work with columns:

In [6]:
vdem = vdem.rename({'country_text_id': 'country_code',
                    'country_name': 'country_name_vdem',
                    'v2x_polyarchy': 'democracy',
                    'v2peedueq':'educational_equality'}, axis=1)
vdem

country_code country_name_vdem  year  democracy  educational_equality
171            MEX            Mexico  1960      0.232                -1.438
172            MEX            Mexico  1961      0.234                -1.438
173            MEX            Mexico  1962      0.233                -1.438
174            MEX            Mexico  1963      0.233                -1.438
175            MEX            Mexico  1964      0.231                -1.438
...            ...               ...   ...        ...                   ...
26150          ZZB          Zanzibar  2017      0.267                 1.661
26151          ZZB          Zanzibar  2018      0.268                 1.486
26152          ZZB          Zanzibar  2019      0.266                 1.486
26153          ZZB          Zanzibar  2020      0.258                 1.427
26154          ZZB          Zanzibar  2021      0.276                 1.779

[10371 rows x 5 columns]

### Part d
Sort the rows by 'country_code' and 'year' in ascending order. [1 point]

To sort, I am using the `.sort_values()` method. Within the method, I use the `by` argument to specify which columns I want to sort by in a list, starting with `country_code`, then `year`. Then I'll pass a list of boolean values to the `ascending` argument to specify whether each of the columns should be sorted in ascending or descending order. In this case both are sorted in ascending order:

In [7]:
vdem = vdem.sort_values(by = ['country_code', 'year'], ascending = [True, True])
vdem

country_code country_name_vdem  year  democracy  educational_equality
5433           AFG       Afghanistan  1960      0.080                -1.123
5434           AFG       Afghanistan  1961      0.083                -1.123
5435           AFG       Afghanistan  1962      0.082                -1.123
5436           AFG       Afghanistan  1963      0.085                -1.123
5437           AFG       Afghanistan  1964      0.137                -0.951
...            ...               ...   ...        ...                   ...
26150          ZZB          Zanzibar  2017      0.267                 1.661
26151          ZZB          Zanzibar  2018      0.268                 1.486
26152          ZZB          Zanzibar  2019      0.266                 1.486
26153          ZZB          Zanzibar  2020      0.258                 1.427
26154          ZZB          Zanzibar  2021      0.276                 1.779

[10371 rows x 5 columns]

## Problem 2
Next focus on the World Bank `wb` dataset 'ESGData.csv'. Use `pandas` methods to perform the following tasks:

### Part a
Keep only the columns named 'Country Code', 'Country Name', and 'Indicator Code', or begin with '19' or '20'. (Don't type in all the years individually. Instead, use code that finds all columns that begin '19' or '20'.) [1 point]

First, I am defining a list `mycols` containing the column names we want to keep:

In [8]:
mycols = ['Country Code', 'Country Name', 'Indicator Code']

Here, I am creating a list of column names from the dataframe, `wb`, that start with '19' or '20'. I am using a list comprehension to iterate over the column names and check if each name starts with '19' or '20'. The resulting list, `yearcols`, contains only those column names that meet this condition:

In [9]:
# getting the year columns
yearcols = [x for x in wb.columns if x.startswith("19") or x.startswith("20")]

Here, I am overwriting the `wb` dataframe with a dataframe only containing `mycols` and `yearcols`:

In [10]:
wb = wb[mycols + yearcols]
wb

Country Code Country Name     Indicator Code  1960       1961  \
0              ARB   Arab World     EG.CFT.ACCS.ZS   NaN        NaN   
1              ARB   Arab World     EG.ELC.ACCS.ZS   NaN        NaN   
2              ARB   Arab World  NY.ADJ.DRES.GN.ZS   NaN        NaN   
3              ARB   Arab World  NY.ADJ.DFOR.GN.ZS   NaN        NaN   
4              ARB   Arab World     AG.LND.AGRI.ZS   NaN  30.981414   
...            ...          ...                ...   ...        ...   
16964          ZWE     Zimbabwe     ER.PTD.TOTL.ZS   NaN        NaN   
16965          ZWE     Zimbabwe     AG.LND.FRLS.HA   NaN        NaN   
16966          ZWE     Zimbabwe     SL.UEM.TOTL.ZS   NaN        NaN   
16967          ZWE     Zimbabwe        SP.UWT.TFRT   NaN        NaN   
16968          ZWE     Zimbabwe             VA.EST   NaN        NaN   

            1962       1963       1964       1965     1966       1967  \
0            NaN        NaN        NaN        NaN      NaN        NaN   
1            NaN        NaN        NaN        NaN      NaN        NaN   
2            NaN        NaN        NaN        NaN      NaN        NaN   
3            NaN        NaN        NaN        NaN      NaN        NaN   
4      30.982663  31.007054  31.018001  31.042466  31.0504  31.103223   
...          ...        ...        ...        ...      ...        ...   
16964        NaN        NaN        NaN        NaN      NaN        NaN   
16965        NaN        NaN        NaN        NaN      NaN        NaN   
16966        NaN        NaN        NaN        NaN      NaN        NaN   
16967        NaN        NaN        NaN        NaN      NaN        NaN   
16968        NaN        NaN        NaN        NaN      NaN        NaN   

            1968       1969       1970       1971       1972       1973  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2            NaN        NaN   5.961207   6.065469   8.242519  11.302722   
3            NaN        NaN   0.174568   0.139287   0.132084   0.152424   
4      31.133565  31.190429  31.254493  31.386588  31.499948  31.496808   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1974       1975       1976       1977       1978       1979  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2      24.108772  16.683838  18.461057  18.377966  15.884196  33.643670   
3       0.087390   0.098760   0.066341   0.102411   0.107103   0.074228   
4      31.550807  31.529648  31.599736  31.621997  31.666078  31.678494   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1980       1981       1982       1983       1984       1985  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2      27.427013  20.161788  11.285801   9.753678   9.670800   8.289274   
3       0.057760   0.055111   0.114144   0.081489   0.079541   0.038306   
4      31.758868  31.454314  31.480303  31.528747  31.942270  32.442177   
...          ...        

### Part b
Rename 'Country Code' to'country_code', 'Country Name' to 'country_name_wb', and 'Indicator Code' to 'feature'. [1 point]

Using the `.rename()` method on the dataframe, `wb`, with a dictionary parameter that contains mapping of the old column names to the new names, and `axis=1` to work with columns:

In [11]:
wb = wb.rename({'Country Code': 'country_code',
                'Country Name': 'country_name_wb',
                'Indicator Code': 'feature'}, axis=1)
wb

country_code country_name_wb            feature  1960       1961  \
0              ARB      Arab World     EG.CFT.ACCS.ZS   NaN        NaN   
1              ARB      Arab World     EG.ELC.ACCS.ZS   NaN        NaN   
2              ARB      Arab World  NY.ADJ.DRES.GN.ZS   NaN        NaN   
3              ARB      Arab World  NY.ADJ.DFOR.GN.ZS   NaN        NaN   
4              ARB      Arab World     AG.LND.AGRI.ZS   NaN  30.981414   
...            ...             ...                ...   ...        ...   
16964          ZWE        Zimbabwe     ER.PTD.TOTL.ZS   NaN        NaN   
16965          ZWE        Zimbabwe     AG.LND.FRLS.HA   NaN        NaN   
16966          ZWE        Zimbabwe     SL.UEM.TOTL.ZS   NaN        NaN   
16967          ZWE        Zimbabwe        SP.UWT.TFRT   NaN        NaN   
16968          ZWE        Zimbabwe             VA.EST   NaN        NaN   

            1962       1963       1964       1965     1966       1967  \
0            NaN        NaN        NaN        NaN      NaN        NaN   
1            NaN        NaN        NaN        NaN      NaN        NaN   
2            NaN        NaN        NaN        NaN      NaN        NaN   
3            NaN        NaN        NaN        NaN      NaN        NaN   
4      30.982663  31.007054  31.018001  31.042466  31.0504  31.103223   
...          ...        ...        ...        ...      ...        ...   
16964        NaN        NaN        NaN        NaN      NaN        NaN   
16965        NaN        NaN        NaN        NaN      NaN        NaN   
16966        NaN        NaN        NaN        NaN      NaN        NaN   
16967        NaN        NaN        NaN        NaN      NaN        NaN   
16968        NaN        NaN        NaN        NaN      NaN        NaN   

            1968       1969       1970       1971       1972       1973  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2            NaN        NaN   5.961207   6.065469   8.242519  11.302722   
3            NaN        NaN   0.174568   0.139287   0.132084   0.152424   
4      31.133565  31.190429  31.254493  31.386588  31.499948  31.496808   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1974       1975       1976       1977       1978       1979  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2      24.108772  16.683838  18.461057  18.377966  15.884196  33.643670   
3       0.087390   0.098760   0.066341   0.102411   0.107103   0.074228   
4      31.550807  31.529648  31.599736  31.621997  31.666078  31.678494   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1980       1981       1982       1983       1984       1985  \
0            NaN        NaN        NaN        NaN        NaN        NaN   
1            NaN        NaN        NaN        NaN        NaN        NaN   
2      27.427013  20.161788  11.285801   9.753678   9.670800   8.289274   
3       0.057760   0.055111   0.114144   0.081489   0.079541   0.038306   
4      31.758868  31.454314  31.480303  31.528747  31.942270  3

### Part c
Use the `.query()` method to remove the rows in which 'country_name_wb' is equal to one of the entries in the folowing `noncountries` list: [1 point]

In [12]:
noncountries = ["Arab World", "Central Europe and the Baltics",
                "Caribbean small states",
                "East Asia & Pacific (excluding high income)",
                "Early-demographic dividend","East Asia & Pacific",
                "Europe & Central Asia (excluding high income)",
                "Europe & Central Asia", "Euro area",
                "European Union","Fragile and conflict affected situations",
                "High income",
                "Heavily indebted poor countries (HIPC)","IBRD only",
                "IDA & IBRD total",
                "IDA total","IDA blend","IDA only",
                "Latin America & Caribbean (excluding high income)",
                "Latin America & Caribbean",
                "Least developed countries: UN classification",
                "Low income","Lower middle income","Low & middle income",
                "Late-demographic dividend","Middle East & North Africa",
                "Middle income",
                "Middle East & North Africa (excluding high income)",
                "North America","OECD members",
                "Other small states","Pre-demographic dividend",
                "Pacific island small states",
                "Post-demographic dividend",
                "Sub-Saharan Africa (excluding high income)",
                "Sub-Saharan Africa",
                "Small states","East Asia & Pacific (IDA & IBRD)",
                "Europe & Central Asia (IDA & IBRD)",
                "Latin America & Caribbean (IDA & IBRD)",
                "Middle East & North Africa (IDA & IBRD)","South Asia",
                "South Asia (IDA & IBRD)",
                "Sub-Saharan Africa (IDA & IBRD)",
                "Upper middle income", "World"]

To remove the rows in which 'country_name_wb' is equal to one of the entries in the `noncountries` list using the `.query()` I am using the following query string: `"country_name_wb not in @noncountries"`.

The query string within the `.query()` call on the dataframe contains `country_name_wb` which refers to the column in the DataFrame that contains country names. `not in`, is a logical operator that checks if the value in `country_name_wb` is not present in the specified list. `@noncountries`, `@` symbol is used to reference a variable that exists outside the query string. In this case, it refers to the `noncountries` list defined earlier.

In [13]:
wb = wb.query("country_name_wb not in @noncountries")
wb

country_code country_name_wb            feature  1960       1961  \
3266           AFG     Afghanistan     EG.CFT.ACCS.ZS   NaN        NaN   
3267           AFG     Afghanistan     EG.ELC.ACCS.ZS   NaN        NaN   
3268           AFG     Afghanistan  NY.ADJ.DRES.GN.ZS   NaN        NaN   
3269           AFG     Afghanistan  NY.ADJ.DFOR.GN.ZS   NaN        NaN   
3270           AFG     Afghanistan     AG.LND.AGRI.ZS   NaN  57.878356   
...            ...             ...                ...   ...        ...   
16964          ZWE        Zimbabwe     ER.PTD.TOTL.ZS   NaN        NaN   
16965          ZWE        Zimbabwe     AG.LND.FRLS.HA   NaN        NaN   
16966          ZWE        Zimbabwe     SL.UEM.TOTL.ZS   NaN        NaN   
16967          ZWE        Zimbabwe        SP.UWT.TFRT   NaN        NaN   
16968          ZWE        Zimbabwe             VA.EST   NaN        NaN   

            1962       1963       1964       1965       1966       1967  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268         NaN        NaN        NaN        NaN        NaN        NaN   
3269         NaN        NaN        NaN        NaN        NaN        NaN   
3270   57.955016  58.031676  58.116002  58.123668  58.192662  58.229459   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1968       1969       1970       1971       1972       1973  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268         NaN        NaN   0.503854   0.644591   0.786744   1.242755   
3269         NaN        NaN   0.279412   0.337567   0.389290   0.751733   
3270   58.230992  58.255523  58.270855  58.316851  58.335250  58.337090   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1974       1975       1976       1977       1978       1979  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268    1.493057   1.849701   1.962964   1.840750   1.618266   1.692939   
3269    0.783186   0.795687   0.756951   0.589740   0.734971   0.559937   
3270   58.338776  58.338776  58.338316  58.338316  58.338316  58.336783   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1980       1981       1982       1983       1984       1985  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268    1.611075   1.226914        NaN        NaN        NaN        NaN   
3269    0.581685   0.434263        NaN        NaN        NaN        NaN   
3270   58.336783  58.342916  58.344449 

### Part d
The features in this dataset are given strange and incomprehensible codes such as 'EG.CFT.ACCS.ZS'. Use the `replace_map` dictionary, defined below, to recode all of these values with more descriptive names for each feature. [1 point]

In [14]:
replace_map = {
  "AG.LND.AGRI.ZS": "agricultural_land",
  "AG.LND.FRST.ZS": "forest_area",
  "AG.PRD.FOOD.XD": "food_production_index",
  "CC.EST": "control_of_corruption",
  "EG.CFT.ACCS.ZS": "access_to_clean_fuels_and_technologies_for_cooking",
  "EG.EGY.PRIM.PP.KD": "energy_intensity_level_of_primary_energy",
  "EG.ELC.ACCS.ZS": "access_to_electricity",
  "EG.ELC.COAL.ZS": "electricity_production_from_coal_sources",
  "EG.ELC.RNEW.ZS": "renewable_electricity_output",
  "EG.FEC.RNEW.ZS": "renewable_energy_consumption",
  "EG.IMP.CONS.ZS": "energy_imports",
  "EG.USE.COMM.FO.ZS": "fossil_fuel_energy_consumption",
  "EG.USE.PCAP.KG.OE": "energy_use",
  "EN.ATM.CO2E.PC": "co2_emissions",
  "EN.ATM.METH.PC": "methane_emissions",
  "EN.ATM.NOXE.PC": "nitrous_oxide_emissions",
  "EN.ATM.PM25.MC.M3": "pm2_5_air_pollution",
  "EN.CLC.CDDY.XD": "cooling_degree_days",
  "EN.CLC.GHGR.MT.CE": "ghg_net_emissions",
  "EN.CLC.HEAT.XD": "heat_index_35",
  "EN.CLC.MDAT.ZS": "droughts",
  "EN.CLC.PRCP.XD": "maximum_5-day_rainfall",
  "EN.CLC.SPEI.XD": "mean_drought_index",
  "EN.MAM.THRD.NO": "mammal_species",
  "EN.POP.DNST": "population_density",
  "ER.H2O.FWTL.ZS": "annual_freshwater_withdrawals",
  "ER.PTD.TOTL.ZS": "terrestrial_and_marine_protected_areas",
  "GB.XPD.RSDV.GD.ZS": "research_and_development_expenditure",
  "GE.EST": "government_effectiveness",
  "IC.BUS.EASE.XQ": "ease_of_doing_business_rank",
  "IC.LGL.CRED.XQ": "strength_of_legal_rights_index",
  "IP.JRN.ARTC.SC": "scientific_and_technical_journal_articles",
  "IP.PAT.RESD": "patent_applications",
  "IT.NET.USER.ZS": "individuals_using_the_internet",
  "NV.AGR.TOTL.ZS": "agriculture",
  "NY.ADJ.DFOR.GN.ZS": "net_forest_depletion",
  "NY.ADJ.DRES.GN.ZS": "natural_resources_depletion",
  "NY.GDP.MKTP.KD.ZG": "gdp_growth",
  "PV.EST": "political_stability_and_absence_of_violence",
  "RL.EST": "rule_of_law",
  "RQ.EST": "regulatory_quality",
  "SE.ADT.LITR.ZS": "literacy_rate",
  "SE.ENR.PRSC.FM.ZS": "gross_school_enrollment",
  "SE.PRM.ENRR": "primary_school_enrollment",
  "SE.XPD.TOTL.GB.ZS": "government_expenditure_on_education",
  "SG.GEN.PARL.ZS": "proportion_of_seats_held_by_women_in_national_parliaments",
  "SH.DTH.COMM.ZS": "cause_of_death",
  "SH.DYN.MORT": "mortality_rate",
  "SH.H2O.SMDW.ZS": "people_using_safely_managed_drinking_water_services",
  "SH.MED.BEDS.ZS": "hospital_beds",
  "SH.STA.OWAD.ZS": "prevalence_of_overweight",
  "SH.STA.SMSS.ZS": "people_using_safely_managed_sanitation_services",
  "SI.DST.FRST.20": "income_share_held_by_lowest_20pct",
  "SI.POV.GINI": "gini_index",
  "SI.POV.NAHC": "poverty_headcount_ratio_at_national_poverty_lines",
  "SI.SPR.PCAP.ZG": "annualized_average_growth_rate_in_per_capita_real_survey_mean_consumption_or_income",
  "SL.TLF.0714.ZS": "children_in_employment",
  "SL.TLF.ACTI.ZS": "labor_force_participation_rate",
  "SL.TLF.CACT.FM.ZS": "ratio_of_female_to_male_labor_force_participation_rate",
  "SL.UEM.TOTL.ZS": "unemployment",
  "SM.POP.NETM": "net_migration",
  "SN.ITK.DEFC.ZS": "prevalence_of_undernourishment",
  "SP.DYN.LE00.IN": "life_expectancy_at_birth",
  "SP.DYN.TFRT.IN": "fertility_rate",
  "SP.POP.65UP.TO.ZS": "population_ages_65_and_above",
  "SP.UWT.TFRT": "unmet_need_for_contraception",
  "VA.EST": "voice_and_accountability",
  "EN.CLC.CSTP.ZS": "coastal_protection",
  "SD.ESR.PERF.XQ": "economic_and_social_rights_performance_score",
  "EN.CLC.HDDY.XD": "heating_degree_days",
  "EN.LND.LTMP.DC": "land_surface_temperature",
  "ER.H2O.FWST.ZS": "freshwater_withdrawal",
  "EN.H2O.BDYS.ZS": "water_quality",
  "AG.LND.FRLS.HA": "tree_cover_loss",
}

Using `.map()` on the `feature` column to apply the mapping defined by the dictionary `replace_map`:

In [15]:
wb.feature = wb.feature.map(replace_map)
wb

/var/folders/c9/9dpn8j8s2cb5mjl9rjm_s0k00000gn/T/ipykernel_10877/215342572.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wb.feature = wb.feature.map(replace_map)


country_code country_name_wb  \
3266           AFG     Afghanistan   
3267           AFG     Afghanistan   
3268           AFG     Afghanistan   
3269           AFG     Afghanistan   
3270           AFG     Afghanistan   
...            ...             ...   
16964          ZWE        Zimbabwe   
16965          ZWE        Zimbabwe   
16966          ZWE        Zimbabwe   
16967          ZWE        Zimbabwe   
16968          ZWE        Zimbabwe   

                                                 feature  1960       1961  \
3266   access_to_clean_fuels_and_technologies_for_coo...   NaN        NaN   
3267                               access_to_electricity   NaN        NaN   
3268                         natural_resources_depletion   NaN        NaN   
3269                                net_forest_depletion   NaN        NaN   
3270                                   agricultural_land   NaN  57.878356   
...                                                  ...   ...        ...   
16964             terrestrial_and_marine_protected_areas   NaN        NaN   
16965                                    tree_cover_loss   NaN        NaN   
16966                                       unemployment   NaN        NaN   
16967                       unmet_need_for_contraception   NaN        NaN   
16968                           voice_and_accountability   NaN        NaN   

            1962       1963       1964       1965       1966       1967  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268         NaN        NaN        NaN        NaN        NaN        NaN   
3269         NaN        NaN        NaN        NaN        NaN        NaN   
3270   57.955016  58.031676  58.116002  58.123668  58.192662  58.229459   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1968       1969       1970       1971       1972       1973  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268         NaN        NaN   0.503854   0.644591   0.786744   1.242755   
3269         NaN        NaN   0.279412   0.337567   0.389290   0.751733   
3270   58.230992  58.255523  58.270855  58.316851  58.335250  58.337090   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN   
16968        NaN        NaN        NaN        NaN        NaN        NaN   

            1974       1975       1976       1977       1978       1979  \
3266         NaN        NaN        NaN        NaN        NaN        NaN   
3267         NaN        NaN        NaN        NaN        NaN        NaN   
3268    1.493057   1.849701   1.962964   1.840750   1.618266   1.692939   
3269    0.783186   0.795687   0.756951   0.589740   0.734971   0.559937   
3270   58.338776  58.338776  58.338316  58.338316  58.338316  58.336783   
...          ...        ...        ...        ...        ...        ...   
16964        NaN        NaN        NaN        NaN        NaN        NaN   
16965        NaN        NaN        NaN        NaN        NaN        NaN   
16966        NaN        NaN        NaN        NaN        NaN        NaN   
16967        NaN        NaN        NaN        NaN        NaN        NaN 

## Problem 3
The `wb` dataset is strangely organized. The features are stored in the rows, when typically we would want these features to be columns. Also, years are stored in columns, when typically we would want years to be represented by different rows. We can repair this structure by reshaping the data. 

### Part a
First, reshape the data to turn the columns that refer to years into rows. [1 point]

To turn the years columns into rows, I will use `pd.melt()`. 
* The first argument in the function is the name of the dataframe, `wb`. 
* The second argument, `id_vars`, which is a list of the columns we want to continue to exist as columns after the dataframe is melted. In our case, it's a list of the following columns: country_code, country_name_wb, and feature.
* The third argument, `value_vars` is a list of the column names we want to store in rows. In our case, these columns are numeric and can be sequentially ordered so I will use `[str(i) for i in range(1960,2022)]`, which loops over integers from 1960 to 2021, placing each year as a string in a list

In [16]:
wb = pd.melt(wb, id_vars = ['country_code', 'country_name_wb', 'feature'],
            value_vars = [str(i) for i in range(1960,2022)])
wb

country_code country_name_wb  \
0               AFG     Afghanistan   
1               AFG     Afghanistan   
2               AFG     Afghanistan   
3               AFG     Afghanistan   
4               AFG     Afghanistan   
...             ...             ...   
849581          ZWE        Zimbabwe   
849582          ZWE        Zimbabwe   
849583          ZWE        Zimbabwe   
849584          ZWE        Zimbabwe   
849585          ZWE        Zimbabwe   

                                                  feature variable      value  
0       access_to_clean_fuels_and_technologies_for_coo...     1960        NaN  
1                                   access_to_electricity     1960        NaN  
2                             natural_resources_depletion     1960        NaN  
3                                    net_forest_depletion     1960        NaN  
4                                       agricultural_land     1960        NaN  
...                                                   ...      ...        ...  
849581             terrestrial_and_marine_protected_areas     2021  27.214747  
849582                                    tree_cover_loss     2021        NaN  
849583                                       unemployment     2021   9.540000  
849584                       unmet_need_for_contraception     2021        NaN  
849585                           voice_and_accountability     2021  -1.135830  

[849586 rows x 5 columns]

### Part b
Then rename `variable` to `year`, and reshape the data again by turning the rows that refer to features into columns. [1 point]

Now the old column names (1960 through 2021 in this case) is stored in a column named “variable” after melting, and the datapoints that populated those columns are now contained in a column named “value”. So, I'm renaming the "variable" columns to 'year':

In [17]:
wb = wb.rename({'variable': 'year'}, axis=1)
wb

country_code country_name_wb  \
0               AFG     Afghanistan   
1               AFG     Afghanistan   
2               AFG     Afghanistan   
3               AFG     Afghanistan   
4               AFG     Afghanistan   
...             ...             ...   
849581          ZWE        Zimbabwe   
849582          ZWE        Zimbabwe   
849583          ZWE        Zimbabwe   
849584          ZWE        Zimbabwe   
849585          ZWE        Zimbabwe   

                                                  feature  year      value  
0       access_to_clean_fuels_and_technologies_for_coo...  1960        NaN  
1                                   access_to_electricity  1960        NaN  
2                             natural_resources_depletion  1960        NaN  
3                                    net_forest_depletion  1960        NaN  
4                                       agricultural_land  1960        NaN  
...                                                   ...   ...        ...  
849581             terrestrial_and_marine_protected_areas  2021  27.214747  
849582                                    tree_cover_loss  2021        NaN  
849583                                       unemployment  2021   9.540000  
849584                       unmet_need_for_contraception  2021        NaN  
849585                           voice_and_accountability  2021  -1.135830  

[849586 rows x 5 columns]

To turn the 'feature' rows into columns, I will use `.pivot_table()` method on the `wb` dataframe. This methods arguments are as follows:
* `index` - a list containing the names of the current columns that we want to remain columns in the reshaped data. In our case, it will be a list of the following columns: country_code, country_name_wb, and year.
* `columns` - the name of the column that contains the names of the new columns we are trying to create. Which is `feature` in our case.
* `values` - the name of the column that contains the datapoints we are trying to move to the new columns.

First, I want to make sure that the `value` column is numeric

In [18]:
wb.dtypes

country_code        object
country_name_wb     object
feature             object
year                object
value              float64
dtype: object

The `value` column is numeric so we can move on to use `.pivot_table()` and chaining `.reset_index()` to ensure that the multi-index created by the pviot is flattened back into regular columns:

In [19]:
wb = wb.pivot_table(index=['country_code', 'country_name_wb', 'year'],
                   columns='feature',
                   values='value').reset_index()
wb

feature country_code country_name_wb  year  \
0                AFG     Afghanistan  1960   
1                AFG     Afghanistan  1961   
2                AFG     Afghanistan  1962   
3                AFG     Afghanistan  1963   
4                AFG     Afghanistan  1964   
...              ...             ...   ...   
11961            ZWE        Zimbabwe  2017   
11962            ZWE        Zimbabwe  2018   
11963            ZWE        Zimbabwe  2019   
11964            ZWE        Zimbabwe  2020   
11965            ZWE        Zimbabwe  2021   

feature  access_to_clean_fuels_and_technologies_for_cooking  \
0                                                      NaN    
1                                                      NaN    
2                                                      NaN    
3                                                      NaN    
4                                                      NaN    
...                                                    ...    
11961                                                 29.8    
11962                                                 30.0    
11963                                                 30.2    
11964                                                 30.3    
11965                                                 30.3    

feature  access_to_electricity  agricultural_land  agriculture  \
0                          NaN                NaN          NaN   
1                          NaN          57.878356          NaN   
2                          NaN          57.955016          NaN   
3                          NaN          58.031676          NaN   
4                          NaN          58.116002          NaN   
...                        ...                ...          ...   
11961                43.979065          41.876696     8.340969   
11962                45.400288          41.876696     7.319375   
11963                46.682095          41.876696     9.819262   
11964                52.747667          41.876696     8.772859   
11965                48.979927          41.876696     8.849899   

feature  annual_freshwater_withdrawals  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   
...                                ...   
11961                        27.234910   
11962                        30.761677   
11963                        30.761677   
11964                        30.761677   
11965                              NaN   

feature  annualized_average_growth_rate_in_per_capita_real_survey_mean_consumption_or_income  \
0                                                      NaN                                     
1                                                      NaN                                     
2                                                      NaN                                     
3                                                      NaN                                     
4                                                      NaN                                     
...                                                    ...                                     
11961                                                  NaN                                     
11962                                                  NaN                                     
11963                                                 1.03                                     
11964                                                  NaN                                     
11965                                                  NaN                                     

feature  cause_of_death  children_in_employment  co2_emissions  \
0                   NaN                     NaN            NaN   
1                   NaN                     NaN            NaN   
2                   NaN                  

### Part c
After these reshapes, the year column in the `wb` data frame is stored as a string. Convert this column to an integer data type. [1 point] 

Using the `.astype()` method to convert the year column to integer:

In [20]:
wb.year = wb.year.astype(int)

In [21]:
wb.year.dtypes

dtype('int64')

## Problem 4
Next we will merge the `wb` data frame with the `vdem` data frame, matching on the 'country_code' and 'year' columns. 

### Part a
First, write a sentence stating whether you expect this merge to be one-to-one, many-to-one, one-to-many, or many-to-many, and describe your rationale. [1 point]

I expect this merge to be one-to-one. The 'country_code' and 'year' combination should uniquely identify each row in both data frames. Each combination of 'country_code' and 'year' in wb corresponds to a single combination of 'country_code' and 'year' in vdem.

### Part b
Next, merge the two datasets together in a way that checks whether your expectation is met, and also allows you to see the rows that failed to match. [2 points]

Using `pd.merge()` to perform an outer merge on 'country_code' and 'year' of the two dataframes, specifiying the keys on which to match rows as a list with the `on` argument, and specifying the type of the merge with the `how` argument. This code also validates that the merge should be **one-to-one** (from wb to vdem), and includes an `indicator` column to show which rows matched and which did not. This approach allows us to verify the type of merge and identify unmatched rows.

In [22]:
merged_data = pd.merge(wb, vdem, 
               on= ['country_code', 'year'], 
               how = 'outer',
               validate = 'one_to_one',
               indicator = 'matched')
merged_data

country_code country_name_wb  year  \
0              AFG     Afghanistan  1960   
1              AFG     Afghanistan  1961   
2              AFG     Afghanistan  1962   
3              AFG     Afghanistan  1963   
4              AFG     Afghanistan  1964   
...            ...             ...   ...   
12356          ZZB             NaN  2017   
12357          ZZB             NaN  2018   
12358          ZZB             NaN  2019   
12359          ZZB             NaN  2020   
12360          ZZB             NaN  2021   

       access_to_clean_fuels_and_technologies_for_cooking  \
0                                                    NaN    
1                                                    NaN    
2                                                    NaN    
3                                                    NaN    
4                                                    NaN    
...                                                  ...    
12356                                                NaN    
12357                                                NaN    
12358                                                NaN    
12359                                                NaN    
12360                                                NaN    

       access_to_electricity  agricultural_land  agriculture  \
0                        NaN                NaN          NaN   
1                        NaN          57.878356          NaN   
2                        NaN          57.955016          NaN   
3                        NaN          58.031676          NaN   
4                        NaN          58.116002          NaN   
...                      ...                ...          ...   
12356                    NaN                NaN          NaN   
12357                    NaN                NaN          NaN   
12358                    NaN                NaN          NaN   
12359                    NaN                NaN          NaN   
12360                    NaN                NaN          NaN   

       annual_freshwater_withdrawals  \
0                                NaN   
1                                NaN   
2                                NaN   
3                                NaN   
4                                NaN   
...                              ...   
12356                            NaN   
12357                            NaN   
12358                            NaN   
12359                            NaN   
12360                            NaN   

       annualized_average_growth_rate_in_per_capita_real_survey_mean_consumption_or_income  \
0                                                    NaN                                     
1                                                    NaN                                     
2                                                    NaN                                     
3                                                    NaN                                     
4                                                    NaN                                     
...                                                  ...                                     
12356                                                NaN                                     
12357                                                NaN                                     
12358                                                NaN                                     
12359                                                NaN                                     
12360                                                NaN                                     

       cause_of_death  children_in_employment  co2_emissions  \
0                 NaN                     NaN            NaN   
1                 NaN                     NaN            NaN   
2                 NaN                     NaN            NaN   
3                 NaN                     NaN            NaN   
4                 NaN                     NaN

### Part c
After this merge, use the `.value_counts()` method to see the total number of observations that were found in both datasets, the number found only in the left dataset, and the number found only in the right dataset. (If you entered the `wb` data frame into the merge function first, then "left_only" refers to the rows found in the World Bank but not V-Dem, and "right_only" refers to the rows found in V-Dem but not the World Bank.) There should be more than 9000 rows that matched, but more than 2000 that failed to match.

Then conduct two data aggregations to help us investigate why these observations did not match:

* First use `.query()` to keep only the observations that were present in `wb` but not `vdem`. (These are the 'left_only' observations if you typed the World Bank data into the merge function first.) Use `.groupby()` to aggregate the data by both 'country_code' and 'country_name_wb'. Then save the minimum and maximum values of 'year' for each country.

* Then use `.query()` to keep only the observations that were present in `vdem` data but not `wb`. Use `.groupby()` to aggregate the data by both 'country_code' and 'country_name_vdem'. Then save the minimum and maximum values of 'year' for each country. [2 points]

In [23]:
merged_data['matched'].value_counts()

matched
both          9976
left_only     1990
right_only     395
Name: count, dtype: int64

I believe these results align with the expectation that there should be more than 9000 rows that matched, and more than 2000 rows that failed to match (total of `left_only` and `right_only` is 2385).

- **Rows found in both datasets ("both")**: 9976
- **Rows found only in the left dataset (World Bank, "left_only")**: 1990
- **Rows found only in the right dataset (V-Dem, "right_only")**: 395

Investigating unmatched observations:

Left Only (World Bank):

In [24]:
left_only = merged_data.query("matched=='left_only'")
left_only.groupby(['country_code', 'country_name_wb'])['year'].agg(['min', 'max']).reset_index()

country_code                 country_name_wb   min   max
0           AND                         Andorra  1960  2021
1           ARE            United Arab Emirates  1960  1970
2           ARM                         Armenia  1960  1989
3           ATG             Antigua and Barbuda  1960  2021
4           AZE                      Azerbaijan  1960  1989
5           BGD                      Bangladesh  1960  1970
6           BHS                    Bahamas, The  1960  2021
7           BIH          Bosnia and Herzegovina  1960  1991
8           BLR                         Belarus  1960  1989
9           BLZ                          Belize  1960  2021
10          BRN               Brunei Darussalam  1960  2021
11          CMR                        Cameroon  1960  1960
12          DMA                        Dominica  1960  2021
13          EST                         Estonia  1960  1989
14          FSM           Micronesia, Fed. Sts.  1960  2021
15          GEO                         Georgia  1960  1989
16          GRD                         Grenada  1960  2021
17          HRV                         Croatia  1960  1990
18          KAZ                      Kazakhstan  1960  1989
19          KGZ                 Kyrgyz Republic  1960  1989
20          KIR                        Kiribati  1960  2021
21          KNA             St. Kitts and Nevis  1960  2021
22          LCA                       St. Lucia  1960  2021
23          LIE                   Liechtenstein  1960  2021
24          LTU                       Lithuania  1960  1989
25          LVA                          Latvia  1960  1989
26          MCO                          Monaco  1960  2021
27          MDA                         Moldova  1960  1989
28          MHL                Marshall Islands  1960  2021
29          MKD                 North Macedonia  1960  1990
30          MNE                      Montenegro  1960  1997
31          NRU                           Nauru  1960  2021
32          PLW                           Palau  1960  2021
33          SMR                      San Marino  1960  2021
34          SSD                     South Sudan  1960  2010
35          SVK                 Slovak Republic  1960  1992
36          SVN                        Slovenia  1960  1988
37          TJK                      Tajikistan  1960  1989
38          TKM                    Turkmenistan  1960  1989
39          TON                           Tonga  1960  2021
40          TUV                          Tuvalu  1960  2021
41          UKR                         Ukraine  1960  1989
42          UZB                      Uzbekistan  1960  1989
43          VCT  St. Vincent and the Grenadines  1960  2021
44          WSM                           Samoa  1960  2021

Right Only (V-Dem):

In [25]:
right_only = merged_data.query("matched=='right_only'")
right_only.groupby(['country_code', 'country_name_vdem'])['year'].agg(['min', 'max']).reset_index()

country_code           country_name_vdem   min   max
0          DDR  German Democratic Republic  1960  1990
1          HKG                   Hong Kong  1960  2021
2          PSE         Palestine/West Bank  1967  2021
3          PSG              Palestine/Gaza  1960  2021
4          SML                  Somaliland  1991  2021
5          TWN                      Taiwan  1960  2021
6          VDR         Republic of Vietnam  1960  1975
7          XKX                      Kosovo  1999  2021
8          YMD                 South Yemen  1960  1990
9          ZZB                    Zanzibar  1960  2021

### Part d
Here's where a deep understanding of the data becomes very important. There are two reasons why an observation may fail to match in a merge. One reason is a difference in spelling. Suppose that South Korea (which is also known as the Republic of Korea) is coded as SKO in the World Bank data and ROK in V-Dem. In this case, we should recode one or the other of SKO and ROK so that they match, otherwise we will lose the data on South Korea. But the second reason why observations might fail to match is due to differences in coverage in the data collection strategy: it is possible that a country wasn't included in one data's coverage, or that certain years for that country were not included. For differences in coverage, there's no way to manipulate the data to match, so we are out of luck and we have to either delete these observations or proceed with missing data from one of the data sources.

Take a close look at the two data aggregation tables you generated in part (j), and answer the following questions:

* Do you see any countries that are present in both the unmatched World Bank rows and the unmatched V-Dem rows, but with different spellings?

I don's see any clear cases of countries present in both lists with different spellings. However, DDR, German Democratic Republic (V-Dem) is likely part of DEU (Germany) in the World Bank data. YMD, South Yemen (V-Dem) is likely part of YEM (Yemen) in the World Bank dataset.

* Do some digging on Wikipedia and other sources on the Internet. What do you think is the primary reason why some countries are present in the V-Dem data but not the World Bank? (You don't need to describe the reasoning for every country. Just dig until you see a general pattern and describe it here.)

The reason could be because of these countries' historical status changes (like German Democratic Republic unification with West Germany) and/or complex political status (such as Palestine/West Bank and Palestine/Gaza). These changes affect how data is collected, recognized, and reported by different organizations over the years. We can see in the case of DDR, German Democratic Republic in V-dem the max year was 1990 which is the year or the country's reunification with West Germany.

V-Dem likely aims to provide a comprehensive historical and political analysis, including countries that may not be universally recognized but are relevant for studying democratic processes.

* Do some more digging on Wikipedia and other sources on the Internet. What do you think is the primary reason why some countries are present in the World Bank data but not V-Dem? (You don't need to describe the reasoning for every country. Just dig until you see a general pattern and describe it here.) [1 point]

On the other hand, the World Bank data appears to focus on currently existing, internationally recognized states. This aligns with the World Bank's role as an international financial institution that typically works with officially recognized countries.

### Part e
Once you are convinced that all of the unmatched observations are due to differences in the coverage of the data collection strategies of the World Bank and V-Dem, repeat the merge, dropping all unmatched observations. This time there is no need to validate the type of merge, and no need to define a variable to indicate matching. [1 point]

To merge, droppin all umatched observations, we can perform an inner join, which will keep only the rows that have matching 'country_code' and 'year' in both datasets:

In [32]:
cardb_data = pd.merge(wb, vdem,
                      on=['country_code', 'year'],
                      how='inner')
cardb_data

country_code country_name_wb  year  \
0             AFG     Afghanistan  1960   
1             AFG     Afghanistan  1961   
2             AFG     Afghanistan  1962   
3             AFG     Afghanistan  1963   
4             AFG     Afghanistan  1964   
...           ...             ...   ...   
9971          ZWE        Zimbabwe  2017   
9972          ZWE        Zimbabwe  2018   
9973          ZWE        Zimbabwe  2019   
9974          ZWE        Zimbabwe  2020   
9975          ZWE        Zimbabwe  2021   

      access_to_clean_fuels_and_technologies_for_cooking  \
0                                                   NaN    
1                                                   NaN    
2                                                   NaN    
3                                                   NaN    
4                                                   NaN    
...                                                 ...    
9971                                               29.8    
9972                                               30.0    
9973                                               30.2    
9974                                               30.3    
9975                                               30.3    

      access_to_electricity  agricultural_land  agriculture  \
0                       NaN                NaN          NaN   
1                       NaN          57.878356          NaN   
2                       NaN          57.955016          NaN   
3                       NaN          58.031676          NaN   
4                       NaN          58.116002          NaN   
...                     ...                ...          ...   
9971              43.979065          41.876696     8.340969   
9972              45.400288          41.876696     7.319375   
9973              46.682095          41.876696     9.819262   
9974              52.747667          41.876696     8.772859   
9975              48.979927          41.876696     8.849899   

      annual_freshwater_withdrawals  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   
...                             ...   
9971                      27.234910   
9972                      30.761677   
9973                      30.761677   
9974                      30.761677   
9975                            NaN   

      annualized_average_growth_rate_in_per_capita_real_survey_mean_consumption_or_income  \
0                                                   NaN                                     
1                                                   NaN                                     
2                                                   NaN                                     
3                                                   NaN                                     
4                                                   NaN                                     
...                                                 ...                                     
9971                                                NaN                                     
9972                                                NaN                                     
9973                                               1.03                                     
9974                                                NaN                                     
9975                                                NaN                                     

      cause_of_death  children_in_employment  co2_emissions  \
0                NaN                     NaN            NaN   
1                NaN                     NaN            NaN   
2                NaN                     NaN            NaN   
3                NaN                     NaN            NaN   
4                NaN                     NaN            NaN   
...              ...                     ...  

## Problem 5
Write code using `pandas` that answers the next two questions:

### Part a
Of all countries in the data, which countries have the highest and lowest average levels of democratic quality across the 1960-2022 timespan? [1 point]

Calculating the average levels of democratic quality for each country by first grouping the data by country using `groupby('country_name_wb')` then aggregate the data to calculate the average democratic quality (`.agg({'democracy': 'mean'})`):

In [33]:
avg_democracy = cardb_data.groupby('country_name_wb').agg({'democracy': 'mean'}).reset_index()
avg_democracy

country_name_wb  democracy
0       Afghanistan   0.179484
1           Albania   0.322790
2           Algeria   0.226790
3            Angola   0.133403
4         Argentina   0.587968
..              ...        ...
167   Venezuela, RB   0.595565
168        Viet Nam   0.170306
169     Yemen, Rep.   0.168806
170          Zambia   0.335403
171        Zimbabwe   0.257661

[172 rows x 2 columns]

To find the countries with the highest and lowest average democratic quality, I am using the `idxmax()` and `idxmin()` on `avg_democracy['democracy']` to find the indices of the countries with the highest and lowest average democracy scores:

In [34]:
max_avg_democracy = avg_democracy['democracy'].idxmax()
min_avg_democracy = avg_democracy['democracy'].idxmin()

Using the indices obtained, I retrieved the corresponding country names and their average democracy scores then printed the results:

In [35]:
max_country = avg_democracy.loc[max_avg_democracy, 'country_name_wb']
min_country = avg_democracy.loc[min_avg_democracy, 'country_name_wb']
max_score = avg_democracy.loc[max_avg_democracy, 'democracy']
min_score = avg_democracy.loc[min_avg_democracy, 'democracy']

print(f"Highest average democratic quality: {max_country} ({round(max_score, 3)})")
print(f"Lowest average democratic quality: {min_country} ({round(min_score, 3)})")

Highest average democratic quality: Denmark (0.91)
Lowest average democratic quality: Saudi Arabia (0.015)


### Part b
The 'educational_equality' index compiled by V-Dem measures the extent to which "high quality basic education guaranteed to all, sufficient to enable them to exercise their basic rights as adult citizens." They use a Bayesian scaling method to create a score for each country in each year that ranges roughly from -4 to 4, where low values of the scale mean that
> Provision of high quality basic education is extremely unequal and at least 75 percent (%) of children receive such low-quality education that undermines their ability to exercise their basic rights as adult citizens.

And high values mean that
> Basic education is equal in quality and less than five percent (%) of children receive such low-quality education that probably undermines their ability to exercise their basic rights as adult citizens.

Use the `pd.cut()` method to create a categorical version of 'educational_equality' with five categories, one from -4 to -2 called "extremely unequal", one from -2 to -.5 called "very unequal", one from -.5 to .5 called "somewhat unequal", one from .5 to 1.5 called "relatively equal", and one for values from 1.5 to 4 called "equal". (By default, the `pd.cut()` method sets `right=True`, which means the bins include their rightmost edges, so a value of exactly -2 will fall within the "extremely unequal" bin. Leave this default in place.)

Then aggregate the data to have one row per category of the new categorical version of "educational_equality". Collapse the following features to the mean with each category of "educational_equality":

* 'gini_index': The GINI index measures the amount of economic inequality in a country. The higher the index, the greater the economic disparity between rich and poor.
* 'poverty_headcount_ratio_at_national_poverty_lines': a measure of the proportion of the population living in poverty [1 point]
  

Breaking the values of the `educational_equality` column into five categories using `pd.cut()`:

In [36]:
cardb_data['educational_equality_cat'] = pd.cut(cardb_data.educational_equality, 
         bins=[-4, -2, -.5, .5, 1.5, 4], 
         labels=("extremely unequal", "very unequal", "somewhat unequal", "relatively equal", "equal"))
cardb_data[['educational_equality', 'educational_equality_cat']]

educational_equality educational_equality_cat
0                   -1.123             very unequal
1                   -1.123             very unequal
2                   -1.123             very unequal
3                   -1.123             very unequal
4                   -0.951             very unequal
...                    ...                      ...
9971                 1.490         relatively equal
9972                 0.995         relatively equal
9973                 0.999         relatively equal
9974                 1.674                    equal
9975                 1.674                    equal

[9976 rows x 2 columns]

Then to show the average GINI index and poverty headcount ratio for each category of educational equality:

In [37]:
cardb_data.groupby('educational_equality_cat').agg({
    'gini_index': 'mean',
    'poverty_headcount_ratio_at_national_poverty_lines': 'mean'
}).reset_index()

/var/folders/c9/9dpn8j8s2cb5mjl9rjm_s0k00000gn/T/ipykernel_10877/376616583.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cardb_data.groupby('educational_equality_cat').agg({


educational_equality_cat  gini_index  \
0        extremely unequal   38.846154   
1             very unequal   45.926484   
2         somewhat unequal   43.200442   
3         relatively equal   37.148861   
4                    equal   32.652901   

   poverty_headcount_ratio_at_national_poverty_lines  
0                                          58.160000  
1                                          38.636058  
2                                          24.149123  
3                                          22.548536  
4                                          17.207444